docs/pages/pmd/userdocs/extending/writing_java_rules.md

```diff
@@ -0,0 +1,164 @@
+
+Most base rule classes use a [Visitor pattern](https://sourcemaking.com/design_patterns/visitor)
+to explore the AST.
+
+### Tree traversal
+
+When a rule is applied to a file, it's handed the root of the AST and told
+to traverse all the tree to look for violations. Each rule defines a specific
+`visit` method for each type of node for of the language, which
+by default just visits the children.
+
+So the following rule would traverse the whole tree and do nothing:
+
+```java
+public class MyRule extends AbstractJavaRule {
+    // all methods are default implementations!
+}
+```
+
+Generally, a rule wants to check for only some node types. In our XPath example
+in [Your First Rule](pmd_userdocs_extending_your_first_rule.html),
+we wanted to check for some `VariableDeclaratorId` nodes. That's the XPath name,
+but in Java, you'll get access to the {% jdoc jast::ASTVariableDeclaratorId %}
+full API.
+
+If you want to check for some specific node types, you can override the
+corresponding `visit` method:
+
+```java
+public class MyRule extends AbstractJavaRule {
+
+    @Override
+    public Object visit(ASTVariableDeclaratorId node, Object data) {
+        // This method is called on each node of type ASTVariableDeclaratorId
+        // in the AST
+
+        if (node.getType() == short.class) {
+            // reports a violation at the position of the node
+            // the "data" parameter is a context object handed to by your rule
+            // the message for the violation is the message defined in the rule declaration XML element
+            addViolation(data, node);
+        }
+
+        // this calls back to the default implementation, which recurses further down the subtree
+        return super.visit(node, data);
+    }
+}
+```
+
+The `super.visit(node, data)` call is super common in rule implementations,
+because it makes the traversal continue by visiting all the descendants of the
+current node.
+
+#### Stopping the traversal
+
+Sometimes you have checked all you needed and you're sure that the descendants
+of a node may not contain violations. In that case, you can avoid calling the
+`super` implementation and the traversal will not continue further down. This
+means that your callbacks (`visit` implementations) won't be called on the rest
+of the subtree. The siblings of the current node may be visited
+recursively nevertheless.
+
+#### Economic traversal: the rulechain
+
+If you don't care about the order in which the nodes are traversed (e.g. your
+rule doesn't maintain any state between visits), then you can monumentally
+speed-up your rule by using the **rulechain**.
+
+That mechanism doesn't recurse on all the tree, instead, your rule will only be
+passed the nodes it is interested in. To use the rulechain correctly:
+* Your rule must register those node types by calling {% jdoc core::Rule#addRuleChainVisit(java.lang.Class) %}
+in its constructor.
+* Your visit methods **must not recurse!** In effect, you should call never
```

Comment:
For this case, we actually use a custom pmd rule: https://github.com/pmd/build-tools/blob/8434ea814f17e3b0fe1c56405b6673eb689d9c7c/src/main/resources/net/sourceforge/pmd/pmd-dogfood-config.xml#L583
Maybe it's worth that we ship this rule in the main distro (in an application specific ruleset), so that it can be used when developing rules for PMD...